memoize more type properties

vtjnash opened the pull request:
type properties are accessed heavily by subtyping and intersection, so it's generally worthwhile to memoize them upon construction.

@carnaval I think you had separately observed this during profiling? Currently, I've also left in the old code paths and use it verify that the cached value is correct when using a debug build.

vtjnash committed:
memoize more type properties

has-typevars is called pretty heavily during intersection and subtyping
so it can be a hot-path

also fix jl_new_type_constructor to stop mutating (aka corrupting)
the immutable TypeVars it is given as a parameter

vtjnash committed:
avoid corrupting immutable TypeVar after building a DataType for them

corrupt them before instead :)
when it's less likely someone is looking

fix #12238
fix #16301

carnaval commented:
yes. I've seen jl_has_typevar show up as a hotspot in perf regularly while doing other things. cool :)

JeffBezanson commented on src/jltypes.c:
```diff
@@ -110,14 +120,17 @@ static int jl_has_typevars__(jl_value_t *v, int incl_wildcard, jl_value_t **p, s
         if (elt != v) {
-            if (jl_has_typevars__(elt, incl_wildcard, p, np))
+            if (jl_has_typevars__(elt, incl_wildcard, p, np)) {
+                if (expect > 0) assert(expect);
```
Doesn't this condition make the assert a no-op?

JeffBezanson commented on src/alloc.c:
```diff
@@ -1011,14 +1015,32 @@ JL_DLLEXPORT jl_datatype_t *jl_new_bitstype(jl_value_t *name, jl_datatype_t *sup
+        jl_tvar_t *tv = (jl_tvar_t*)jl_svecref(p, i);
+        assert(jl_is_typevar(tv));
+        env[2 * i] = (jl_value_t*)tv;
+        tv = jl_new_typevar(tv->name, tv->lb, tv->ub);
```
Did you mean `0` instead of `tv->ub`? Otherwise I'm not sure what this is doing. Or we could have the front end pass in unbound typevars in the first place I think.

vtjnash replied:
why would I make a TypeVar with a ub of NULL? Why hasn't the frontend always been doing that instead of requiring the backend to mutate them?

JeffBezanson replied:
oops, I completely misread that. But yes, the frontend should do it. That must be an artifact from before the `TypeVar` constructor exposed this "feature".

vtjnash committed:
fix lowering of TypeVar in TypeConstructor typealias expression

have lowering emit the correct TypeVar expression instead of fixing it
in the TypeConstructor constructor

vtjnash committed:
fix lowering of TypeVars in DataType declations

have lowering emit the correct TypeVar expression
instead of fixing it in the DataType constructor

JeffBezanson commented:
LGTM. Looks like some commits should be squashed.

StefanKarpinski commented:
Fortunately, that's now easy to do in GitHub so you can actually just merge this if you want.
